Prepend control characters to mention text

cyuan1124 opened the pull request:
When the user taps "@Chen" and selects entity "Chen Yuan", currently the highlighted mention text in the text view is "Chen Yuan". This PR is to implement the feature to prepend control character ('@') to mention entity text. Then the text view will display "@Chen Yuan".

The feature is implemented in the following steps:

1. Add a new parameter in HKWMentionsPluginV2's initializer to take in an NSCharacterSet of control characters to be prepended. Also add a class property for default at symbols, including English character (@) and Japanese character (＠).

2. When creating a new MentionsAttribue from the state machine, we check if the mention is triggered by a prepend-able control character, and if so, modify the text in the MentionsAttrirbute to include that control character at the beginning.  With this approach, we don't need to modify `- (NSArray *)mentions` and `- (void)addMention:(HKWMentionsAttribute *)mention` functions.

3. Add unit tests to verify the mention plugin's behaviors on different prepend control character sets, and do manual tests with English/Arabic/Japanese.

Below are the demo videos of English and Arabic.

https://user-images.githubusercontent.com/5710914/112066951-a4daca00-8b3d-11eb-8eaf-d27c9fa2f572.mov

https://user-images.githubusercontent.com/5710914/112066953-a4daca00-8b3d-11eb-8420-13cb9b73af39.mov

Demo for moving cursor:

https://user-images.githubusercontent.com/5710914/112177424-478f5900-8bcf-11eb-89d4-c792c79aec71.mov

bkoatz commented on HakawaiTests/HKWMentionsPluginTests.m:
```diff
@@ -993,6 +996,73 @@ - (CGFloat)heightForCellForMentionsEntity:(id<HKWMentionsEntityProtocol> _Null_u
+        
+        expect([mentionsPlugin mostRecentControlCharacterLocationInText:@"@FirstName LastName test"]).to.equal(NSNotFound);
+    });
+});
```
please add a test for "selectedRange", where u place the cursor in one of these mentions entities with a control char prepended, and make sure the selected range is correct

bkoatz reviewed (changes requested):
looks great minus my small comments + a lix. Could u also manually test this with the | cursor being dragged over and placed with control-char-prepended mentions? along w adding the in-code test. great work

cyuan1124 commented:
Addressed comments:
- Add a flag in HKWTextView to enable pretending control character
- Check flag value in MentionsPluginV2 before attempting to prepend control character
- Add a unit test case to verify highlighted range includes prepended control character

bkoatz commented on Hakawai/HKWExternalMentionConstants.h:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ Character set for at symbols to prepend to mention text, including @, ＠ (Japanese), for the ease of library consumers.
+ */
```
if you could style this as:

Character set for at symbols to prepend to mention text, including @, ＠ (Japanese)
NOTE: This is for the ease of library consumers, not for use inside the library

bkoatz reviewed (changes requested):
last little change, then it's good to me.

cyuan1124 commented:
Updated to fix the comment.
